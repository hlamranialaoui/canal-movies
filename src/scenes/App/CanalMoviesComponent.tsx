import React, {useEffect, useState} from 'react';
import './CanalMovies.css';
import SearchComponent from "./components/SearchComponent";
import MovieComponent from "./components/MovieComponent";
import {IMovie} from "./model/movie";
import InfiniteScroll from "react-infinite-scroll-component";

const defaultProps = {};

export interface IPaginatedMovies {
    movieList: IMovie[];
    totalPages: number;
    currentPage: number;
    totalResults: number;
    query?: string;
}

interface IAppComponentProps {
    trendingMovies: IPaginatedMovies;
    filteredMovies: IPaginatedMovies;
    loadMovieList: Function;
    filterMovieList: Function;
}

const FilteredMoviesView = ({movies, getMoreData}) => (
    <InfiniteScroll
        dataLength={movies.movieList.length}
        next={getMoreData}
        hasMore={movies.totalPages !== movies.currentPage}
        scrollableTarget="body-main"
        style={{overflow: 'none'}}
        endMessage={<p>No further results found</p>}
        loader={<></>}
    >
        <div className="row">
            {movies.movieList?.map((movie, index) =>
                (<MovieComponent
                    key={index}
                    originalName={movie.original_title ? movie.original_title : movie.name}
                    posterPath={movie.poster_path}
                    voteAverage={movie.vote_average}
                />))
            }
        </div>
    </InfiniteScroll>);


const CanalMoviesComponent = ({
                                  trendingMovies,
                                  filteredMovies,
                                  filterMovieList,
                                  loadMovieList
                              }: IAppComponentProps): React.ReactElement => {
    const [query, setQuery] = useState<string>('');

    useEffect(() => {
        loadMovieList(1);
    }, [])

    useEffect(() => {
        if (query) filterMovieList(filteredMovies.currentPage, query);
    }, [query])

    const getMoreData = () => {
        loadMovieList(trendingMovies.currentPage + 1);
    }

    const getMoreDataFiltered = () => {
        filterMovieList(filteredMovies.currentPage + 1, query);
    }

    return (
        <div className="App">
            <nav className="navbar navbar-light p-0">
                <div className="container-fluid px-5">
                    <a className="navbar-brand">CANAL Movies</a>
                </div>
            </nav>
            <SearchComponent filter={setQuery}/>
            <FilteredMoviesView movies={query ? filteredMovies : trendingMovies}
                                getMoreData={query ? getMoreDataFiltered : getMoreData}/>)
        </div>
    );
};

CanalMoviesComponent.defaultProps = defaultProps;

export default CanalMoviesComponent;
