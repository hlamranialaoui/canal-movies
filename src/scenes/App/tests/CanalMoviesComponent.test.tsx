import React from 'react';
import { cleanup, render } from '@testing-library/react';
import CanalMoviesComponent from "../CanalMoviesComponent";
import {FilteredMoviesMock, TrendingMoviesMock} from "./mocks/CanalMoviesComponentMocks";

afterEach(cleanup);

describe('renders canal movies app entry', () => {

    const loadMoviesList = () => TrendingMoviesMock;
    const filteredMoviesList = () => FilteredMoviesMock;

    test('renders movies', async () => {
        const { asFragment } = render(
            <CanalMoviesComponent
                trendingMovies={TrendingMoviesMock}
                filteredMovies={FilteredMoviesMock}
                loadMovieList={loadMoviesList}
                filterMovieList={filteredMoviesList} />
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
