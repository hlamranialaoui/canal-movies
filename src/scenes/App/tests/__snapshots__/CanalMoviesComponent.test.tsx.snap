// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders canal movies app entry renders movies 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <nav
      class="navbar navbar-light p-0"
    >
      <div
        class="container-fluid px-5"
      >
        <a
          class="navbar-brand"
        >
          CANAL Movies
        </a>
      </div>
    </nav>
    <div
      class="bg-img"
    >
      <div
        class="layer"
      >
        <div
          class="row w-100"
        >
          <div
            class="col-md-2 col-1"
          />
          <div
            class="col-md-8 col-8"
          >
            <div
              class="input-group mb-3"
            >
              <input
                class="form-control"
                placeholder="search movie"
                type="text"
              />
              <span
                class="input-group-text"
                id="basic-addon2"
              >
                Search by movie name
              </span>
            </div>
          </div>
          <div
            class="col-md-2 col-1"
          />
        </div>
      </div>
    </div>
    <div
      class="infinite-scroll-component__outerdiv"
    >
      <div
        class="infinite-scroll-component "
        style="height: auto; overflow: none;"
      >
        <div
          class="row"
        >
          <section
            class="col-lg-3"
          >
            <div
              class="container"
            >
              <div
                class="row"
              >
                <div
                  class="col-lg-12"
                >
                  <img
                    alt="single"
                    src="https://image.tmdb.org/t/p/w500/string"
                    style="max-width: 80%;"
                  />
                </div>
              </div>
              <div
                class="row align-items-center p-2"
              >
                <div
                  class="col-lg-12"
                >
                  <h5>
                    ORIGINAL_TITLE
                  </h5>
                  <p
                    class="movie-vote-average"
                  >
                    <i
                      class="bi bi-star-fill"
                    />
                     3
                  </p>
                </div>
              </div>
            </div>
          </section>
          <section
            class="col-lg-3"
          >
            <div
              class="container"
            >
              <div
                class="row"
              >
                <div
                  class="col-lg-12"
                >
                  <img
                    alt="single"
                    src="https://image.tmdb.org/t/p/w500/string"
                    style="max-width: 80%;"
                  />
                </div>
              </div>
              <div
                class="row align-items-center p-2"
              >
                <div
                  class="col-lg-12"
                >
                  <h5>
                    ORIGINAL_TITLE
                  </h5>
                  <p
                    class="movie-vote-average"
                  >
                    <i
                      class="bi bi-star-fill"
                    />
                     3
                  </p>
                </div>
              </div>
            </div>
          </section>
        </div>
        <p>
          No further results found
        </p>
      </div>
    </div>
    )
  </div>
</DocumentFragment>
`;
